### PR TITLE
Editorial edits for chapter 8, and focus cleanup

### PR DESCRIPTION
**Pull request by chrishtr**
* There is a focus field on both Browser and Chrome. The former can be set only to "content" and the latter only to "address book".
* Schedule render on keypress and enter for chapters 12+ (bugfix I found while fixing focus).
* Fixed one broken `self_rect` method.

### CODE REVIEW

File: book/forms.md [hunk 1]
```diff
@@ -398,33 +398,35 @@ remember that in its own `focus` field!
...
 ```
 
-The `if` branch that corresponds to clicks in the browser interface
-unsets `focus` by default, but some existing code in that branch will
-set `focus` to `"address bar"` if the user actually clicked in the
-address bar.
+The `if` branch that corresponds to clicks in the browser chrome
+unsets `focus`, but some existing code in that branch will
+set the `Chrome` `focus` to `"address bar"` if the user actually clicked
+in the address bar.
```
Comment (pavpanchekha) [inline book/forms.md:419]:
```suggestion
The `if` branch that corresponds to clicks in the browser chrome
unsets `focus`, meaning focus is no longer on the page contents,
and key presses will thus be sent to the `Chrome`.
```

File: book/forms.md [hunk 2]
```diff
@@ -398,33 +398,35 @@ remember that in its own `focus` field!
...
-bar or calls the active tab's `keypress` method:
+bar or calls the active tab's `keypress` method (or neither, if nothing is
+focused):
 
 ``` {.python}
 class Browser:
     def handle_key(self, e):
         # ...
-        if self.focus == "chrome":
+        if self.chrome.focus:
```
Comment (pavpanchekha) [inline book/forms.md:429]: I say this should instead be `if self.focus == None`, and then let the `Chrome` sort out whether anything inside the chrome is focused. I'm trying to preserve some notion of hierarchy here.

File: book/scheduling.md [hunk 3]
```diff
@@ -741,12 +741,12 @@ class Browser:
             self.set_needs_raster_and_draw()
 
     def handle_key(self, char):
-        if self.focus == "chrome":
+        if self.chrome.focus:
```
Comment (pavpanchekha) [inline book/scheduling.md:744]: Same `self.focus == None` here.

File: book/scheduling.md [hunk 4]
```diff
@@ -741,12 +741,12 @@ class Browser:
...
             # ...
             self.set_needs_raster_and_draw()
 
     def handle_enter(self):
-        if self.focus == "chrome":
+        if self.chrome.focus:
```
Comment (pavpanchekha) [inline book/scheduling.md:749]: Same `self.focus == None` here.

File: book/scheduling.md [hunk 5]
```diff
@@ -1340,7 +1340,7 @@ The same logic holds for `keypress`:
 class Browser:
     def handle_key(self, char):
         if not (0x20 <= ord(char) < 0x7f): return
-        if self.focus == "chrome":
+        if self.chrome.focus:
```
Comment (pavpanchekha) [inline book/scheduling.md:1343]: Same `self.focus == None` here.

File: src/lab11.py [hunk 6]
```diff
@@ -539,6 +539,7 @@ def click(self, x, y):
             self.browser.active_tab.go_back()
         elif self.address_rect.contains(x, y):
             self.focus = "address bar"
+            print('set focus')
```
Comment (pavpanchekha) [inline src/lab11.py:542]: Woops!

File: src/lab11.py [hunk 7]
```diff
@@ -604,7 +606,7 @@ def handle_click(self, e):
 
     def handle_key(self, char):
         if not (0x20 <= ord(char) < 0x7f): return
-        if self.focus == "chrome":
+        if self.chrome.focus:
```
Comment (pavpanchekha) [inline src/lab11.py:609]: Same `self.focus == None` here.

File: src/lab11.py [hunk 8]
```diff
@@ -587,12 +588,13 @@ def __init__(self):
...
     def handle_click(self, e):
         if e.y < self.chrome.bottom:
-            self.focus = "chrome"
+            self.focus = None
             self.chrome.click(e.x, e.y)
             self.raster_chrome()
         else:
             if self.focus != "content":
                 self.focus = "content"
+                self.chrome.focus = None
```
Comment (pavpanchekha) [inline src/lab11.py:597]: Ooh. Does this actually mean we need to add blur methods to Chrome? Do we correctly handle clicking on the address bar while editing an input element? Or the reverse?

File: src/lab11.py [hunk 9]
```diff
@@ -614,7 +616,7 @@ def handle_key(self, char):
             self.draw()
 
     def handle_enter(self):
-        if self.focus == "chrome":
+        if self.chrome.focus:
```
Comment (pavpanchekha) [inline src/lab11.py:619]: Same `self.focus == None` here.

**Review by pavpanchekha — COMMENTED** [review]
Ok, I think the simplification is good, there's no reason from browser focus to have three states. But I'm still not sure we're handling focus correctly, especially in cases where you are editing the address bar / an input and then click in the other. Can we add tests for this? Also I think it would be neat if we could maintain hierarchy, that is, `Browser` never looks at `Chrome`'s `focus` field. I think that should be possible. It might require adding `blur` methods (maybe we call it `unfocus`) but those might be necessary anyway because I think there might be bugs.

If you'd like, I'm fine merging this and doing the above in a separate PR, or just doing it in this PR.

File: book/forms.md (hunk 2 above)
Comment (chrishtr) [reply book/forms.md:428]: I'll fix this in a followup.

File: src/lab11.py (hunk 8 above)
Comment (chrishtr) [reply src/lab11.py:595]: I tested it pretty carefully and it works. I added a unittest though just now to further ensure correctness.

**Pull request merged**